### PR TITLE
Validate host certificates in both tsh as well as the recording proxy.

### PR DESCRIPTION
**Pull request by russjones**
**Purpose**

At the moment tsh and the recording proxy make sure that host certificates are generated by Teleport. This PR also checks that the host certificate timestamp and hostname (or IP).

**Implementation**

* When generating a host certificate add advertise IP into the host certificate.
* If no advertise IP is specified add `0.0.0.0` to the host certificate and let the Auth Server replace it with the remote address when generating a host certificate.
* Upon dailing, pass in the host address so the forwarding proxy can generate the appropriate host certificate.
* Check that the timestamp and hostname (or IP) of on the host certificate is valid.
* If local registration is occurring add localhost to the host certificate.

### CODE REVIEW

**Comment by russjones** [comment]
@gravitational-jenkins retest this please

File: lib/auth/auth.go [hunk 1]
```diff
@@ -758,6 +758,10 @@ type GenerateServerKeysRequest struct {
 	// if present will be signed as a return value
 	// otherwise, new public/private key pair will be generated
 	PublicSSHKey []byte `json:"public_ssh_key"`
+	// RemoteAddr is the IP address of the remote host host requesting a host
```
Comment (klizhentas) [inline lib/auth/auth.go]: ~~host~~ host

File: lib/auth/auth.go [hunk 2]
```diff
@@ -758,6 +758,10 @@ type GenerateServerKeysRequest struct {
...
+	// certificate. RemoteAddr is used to replace If 0.0.0.0 in the list of
```
Comment (klizhentas) [inline lib/auth/auth.go]: ~~If~~interface

File: lib/client/keyagent.go [hunk 3]
```diff
@@ -258,77 +258,99 @@ func (a *LocalKeyAgent) UserRefusedHosts() bool {
...
 	}
-	// final step: if we have not seen the host key/cert before, lets ask the user if
-	// he trusts it, and add to the known_hosts if he says "yes"
-	if err = hostPromptFunc(host, key); err != nil {
-		// he said "no"
-		a.noHosts[host] = true
+
+	// If this certificate was not seen before, prompt the user essentially
+	// treating it like a key.
+	err = a.checkHostKey(addr, nil, key)
```
Comment (klizhentas) [inline lib/client/keyagent.go]: isn't this redundant with `HostKeyFallback` in place?

File: lib/client/keyagent.go [hunk 4]
```diff
@@ -258,77 +258,99 @@ func (a *LocalKeyAgent) UserRefusedHosts() bool {
...
+		fmt.Printf("The authenticity of host '%s' can't be established. "+
+			"Its public key is:\n%s\nAre you sure you want to continue (yes/no)? ",
+			host, ssh.MarshalAuthorizedKey(key))
+
+		bytes := make([]byte, 12)
+		os.Stdin.Read(bytes)
+		userAnswer = strings.TrimSpace(strings.ToLower(string(bytes)))
 	}
-	return err
+	if !strings.HasPrefix(userAnswer, "y") {
```
Comment (klizhentas) [inline lib/client/keyagent.go]: we have `utils.ParseBool`, can you check if you can use it instead?

File: lib/client/keyagent.go [hunk 5]
```diff
@@ -258,77 +258,99 @@ func (a *LocalKeyAgent) UserRefusedHosts() bool {
...
+		a.log.Warnf("Failed to save the host key: %v.", err)
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+// defaultHostPromptFunc is the default host key/certificates prompt.
+func (a *LocalKeyAgent) defaultHostPromptFunc(host string, key ssh.PublicKey) error {
+	userAnswer := "no"
+	if !a.noHosts[host] {
```
Comment (klizhentas) [inline lib/client/keyagent.go]: I don't understand the rationale behind `a.noHosts` + not sure if we want to keep some local persistent concurrent cache of those? Can you come up with alternative design?

File: lib/client/keyagent.go [hunk 6]
```diff
@@ -258,77 +258,99 @@ func (a *LocalKeyAgent) UserRefusedHosts() bool {
...
+
+		bytes := make([]byte, 12)
+		os.Stdin.Read(bytes)
+		userAnswer = strings.TrimSpace(strings.ToLower(string(bytes)))
 	}
-	return err
+	if !strings.HasPrefix(userAnswer, "y") {
+		return trace.AccessDenied("untrusted host %v", host)
+	}
+	// success
```
Comment (klizhentas) [inline lib/client/keyagent.go]: redundant comment

File: lib/reversetunnel/cache.go [hunk 7]
```diff
@@ -60,25 +59,23 @@ func NewHostCertificateCache(keygen sshca.Authority, authClient auth.ClientI) (*
 // Multiple callers can arrive and generate a host certificate at the same time.
 // This is a tradeoff to prevent long delays here due to the expensive
 // certificate generation call.
-func (c *certificateCache) GetHostCertificate(addr string) (ssh.Signer, error) {
+func (c *certificateCache) GetHostCertificate(addr string, additionalPrincipals []string) (ssh.Signer, error) {
```
Comment (klizhentas) [inline lib/reversetunnel/cache.go]: what if the list of additional principals for the certificate has changed? you might want to invalidate the cache then. Kube forwarder solves this problem by making principals to be part of the key:

https://github.com/gravitational/teleport/blob/master/lib/kube/proxy/forwarder.go#L194

Could something similar be relevant in your use-case as well?

File: lib/service/connect.go [hunk 8]
```diff
@@ -477,6 +477,12 @@ func (process *TeleportProcess) rotate(conn *Connector, localState auth.StateV2,
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
+	additionalPrincipals = utils.ReplaceInSlice(
```
Comment (klizhentas) [inline lib/service/connect.go]: I'm not sure if `0.0.0.0` is a good thing to be present in principals, is there a way around this?

**Review by klizhentas — CHANGES_REQUESTED** [review]

File: lib/client/keyagent.go (hunk 3 above)
Comment (russjones) [reply lib/client/keyagent.go]: This prompts the user to trust a host certificate, in `HostKeyFallback` the user is prompted to check a host key.

File: lib/client/keyagent.go (hunk 5 above)
Comment (russjones) [reply lib/client/keyagent.go]: This was just a straight port of the originally code. Looking into it, it doesn't appear to do anything. I'll remove it.

File: lib/service/connect.go (hunk 8 above)
Comment (russjones) [reply lib/service/connect.go]: `0.0.0.0` is never added to the list of principals, it's always either removed to replaced. This is similar to how to heartbeat logic works, `0.0.0.0` is used to detect that the client does not know what it's IP is.

File: lib/client/keyagent.go (hunk 5 above)
Comment (russjones) [reply lib/client/keyagent.go]: Actually looked at it again, it's used in tests. Updated the comment for `noHosts` to reflect what it does and where it's used (tests).

**Comment by russjones** [comment]
@gravitational-jenkins retest this please

File: lib/client/keyagent.go [hunk 9]
```diff
@@ -258,77 +259,99 @@ func (a *LocalKeyAgent) UserRefusedHosts() bool {
...
+		fmt.Printf("The authenticity of host '%s' can't be established. "+
+			"Its public key is:\n%s\nAre you sure you want to continue (yes/no)? ",
+			host, ssh.MarshalAuthorizedKey(key))
+
+		bytes := make([]byte, 12)
+		os.Stdin.Read(bytes)
+		userAnswer = strings.TrimSpace(strings.ToLower(string(bytes)))
+	}
+	ok, err := utils.ParseBool(userAnswer)
+	if err != nil || !ok {
```
Comment (klizhentas) [inline lib/client/keyagent.go]: I think there is a mistake, `if err == nil and !ok` , you will return `trace.Wrap(nil)` what is the opposite of what we want?

Comment (klizhentas) [reply lib/client/keyagent.go]: If I'm right, can you add the test for this to make sure we won't make this mistake again?

**Review by klizhentas — CHANGES_REQUESTED** [review]
can you check one more comment before merging this in?

**Pull request merged**